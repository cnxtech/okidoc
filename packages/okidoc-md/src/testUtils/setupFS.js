import path from 'path';
import fs from 'fs-extra';
import tempy from 'tempy';

// NOTE: inspired by https://github.com/wix/haste/blob/v0.3.1/packages/haste-test-utils-core/src/fs-setup.js

/**
 * Setup temp directory with test utils
 * @param {Object} fsObject
 *
 * @example
 *
 * const testFS = setupFS({
 *   'src/constants.ts': 'export const X = 1;'
 *   'src/index.ts': 'import { X } from './constants'
 * })
 *
 * console.log(testFS.cwd)
 * // /path/to/temp/directory
 *
 * console.log(testFS.files['src/index.ts'].path)
 * // /path/to/temp/directory/src/index.ts
 */
function setupFS(fsObject = {}) {
  const cwd = tempy.directory();

  Object.keys(fsObject).forEach(fileName => {
    fs.outputFileSync(path.join(cwd, fileName), fsObject[fileName]);
  });

  const createFileProxy = fileName => {
    return new Proxy(
      {},
      {
        get(target, prop) {
          const filePath = path.join(cwd, fileName);

          switch (prop) {
            case 'content':
              return fs.readFileSync(filePath, 'utf8');

            case 'exists':
              return fs.existsSync(filePath);

            case 'path':
              return filePath;

            case 'write':
              return content => fs.outputFileSync(filePath, content);

            default:
              return undefined;
          }
        },
      },
    );
  };

  const files = new Proxy(
    {},
    {
      get: (target, fileName) => createFileProxy(fileName),
    },
  );

  return {
    cwd,
    files,
  };
}

function fromFixture(fixturesDir, fixtureFileName) {
  return fs.readFileSync(path.join(fixturesDir, fixtureFileName), 'utf8');
}

/**
 * @param {string} fixturesDir
 * @param {string[]} fixturesFileNames
 *
 * @example
 *
 * setupFS(fromFixtures('/path/to/fixtures', [
 *  '/src/index.ts',
 *  '/docs.yml'
 * ]))
 */
function fromFixtures(fixturesDir, fixturesFileNames) {
  return fixturesFileNames.reduce((fsObject, fileName) => {
    fsObject[fileName] = fromFixture(fixturesDir, fileName);

    return fsObject;
  }, {});
}

/**
 * @param {string} fixturesDir
 * @param {Object} fixturesMap
 *
 * @example
 *
 * setupFS(fromFixtures('/path/to/fixtures', {
 *   'src/index.ts': 'src/invalid-index.ts',
 *   'docs.yml': 'src/docs.yml'
 * }))
 */
function fromFixturesMap(fixturesDir, fixturesMap) {
  return Object.keys(fixturesMap).reduce((fsObject, fileName) => {
    fsObject[fileName] = fromFixture(fixturesDir, fsObject[fileName]);

    return fsObject;
  }, {});
}

export { fromFixture, fromFixtures, fromFixturesMap };

export default setupFS;
