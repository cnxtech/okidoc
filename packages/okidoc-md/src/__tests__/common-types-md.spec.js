import setupFS, { fromFixtures } from '../testUtils/setupFS';
import runCLI from '../cli';

import path from 'path';

describe('with common types', function() {
  const cwd = process.cwd();

  afterEach(() => {
    process.chdir(cwd);
  });

  it('should render doc markdown for common types', async () => {
    const testFS = setupFS(
      fromFixtures(path.join(__dirname, '/fixtures/common-types-md'), [
        'src/types.ts',
        'src/App.ts',
        'src/AppTwo.ts',
        'docs.yml',
      ]),
    );

    process.chdir(testFS.cwd);

    await runCLI({
      configPath: testFS.files['/docs.yml'].path,
      outputDir: testFS.files['/docs'].path,
    });

    const app = testFS.files['/docs/app.md'].content;
    const types = testFS.files['/docs/types.md'].content;

    expect(app.includes('## IRunResult')).toBeFalsy(); // should not have type definition in doc
    expect(types).toMatchSnapshot();
  });

  it('should use common types config if any', async () => {
    const testFS = setupFS(
      fromFixtures(path.join(__dirname, '/fixtures/common-types-md'), [
        'src/types.ts',
        'src/App.ts',
        'src/AppTwo.ts',
        'docs-with-common-types.yml',
      ]),
    );

    process.chdir(testFS.cwd);

    await runCLI({
      configPath: testFS.files['/docs-with-common-types.yml'].path,
      outputDir: testFS.files['/docs'].path,
    });

    const types = testFS.files['/docs/my-types.md'].content;

    expect(types.includes('# My Types')).toBeTruthy();
    expect(types).toMatchSnapshot();
  });

  it('should render doc markdown for common types inside same md for single page documentation', async () => {
    const testFS = setupFS(
      fromFixtures(path.join(__dirname, '/fixtures/common-types-md'), [
        'src/types.ts',
        'src/App.ts',
        'docs-single.yml',
      ]),
    );

    process.chdir(testFS.cwd);

    await runCLI({
      configPath: testFS.files['/docs-single.yml'].path,
      outputDir: testFS.files['/docs'].path,
    });

    expect(
      testFS.files['/docs/app.md'].content.includes('## IRunResult'),
    ).toBeTruthy(); // should have type definition in doc
    expect(testFS.files['/docs/types.md'].exists).toBeFalsy();
  });

  it('should render types inside same md for multiple page documentation if common types config is not set', async () => {
    const testFS = setupFS(
      fromFixtures(path.join(__dirname, '/fixtures/common-types-md'), [
        'src/types.ts',
        'src/App.ts',
        'src/AppTwo.ts',
        'docs-multiple.yml',
      ]),
    );

    process.chdir(testFS.cwd);

    await runCLI({
      configPath: testFS.files['/docs-multiple.yml'].path,
      outputDir: testFS.files['/docs'].path,
    });

    const app = testFS.files['/docs/app.md'].content;
    const appTwo = testFS.files['/docs/app-two.md'].content;

    expect(app.includes('## IRunResult')).toBeTruthy();
    expect(appTwo.includes('## IRunResult')).toBeTruthy();
    expect(testFS.files['/docs/types.md'].exists).toBeFalsy();
  });
});
