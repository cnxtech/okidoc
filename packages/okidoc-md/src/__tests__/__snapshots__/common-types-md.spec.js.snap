// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with common types should use common types config if any 1`] = `
"<!-- Generated automatically. Update this documentation by updating the source code. -->

# My Types

## IRunResult

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">MEMBERS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>

## IRunResult

<div class=\\"method-list\\">
  <table>
    <thead>
      <tr>
        <th class=\\"title\\">MEMBERS</th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td class=\\"param\\">
          <code>x</code>
        </td>
        <td>
            <div class=\\"type\\">number</div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
"
`;
